File: python/astroengine/rajayoga.py
"""astroengine.rajayoga -- the lordship-and-aspect layer for Vedic yogas, and
the raja/dhana yogas built on it.

Layers, each deterministic:
- House lordship: the traditional ruler of the sign on each whole-sign house
  from the Ascendant.
- Graha drishti (Vedic aspects): every planet aspects the 7th sign from itself;
  Mars also the 4th and 8th, Jupiter the 5th and 9th, Saturn the 3rd and 10th.
- Association: two planets form a yoga when they conjoin, mutually aspect, or
  exchange signs (parivartana).
- Raja yoga: an association between a kendra lord (1/4/7/10) and a trikona lord
  (1/5/9). Dhana yoga: an association between two wealth-house lords (2/5/9/11).
  A yogakaraka is a single planet ruling both a pure kendra (4/7/10) and a pure
  trikona (5/9).

The yoga definitions follow BPHS; the convention (the drishti table, the kendra/
trikona/dhana house sets) is validated against the named source in
`validate_jyotish`, not asserted. The TS port (rajayoga.ts) reproduces every
value and the golden fixtures pin the two together.
"""
from .profections import SIGN_RULERS

# Graha drishti: the house-distances (1-based) each planet aspects.
DRISHTI = {
    "sun": [7], "moon": [7], "mercury": [7], "venus": [7],
    "mars": [4, 7, 8], "jupiter": [5, 7, 9], "saturn": [3, 7, 10],
}
KENDRAS = [1, 4, 7, 10]
TRIKONAS = [1, 5, 9]
DHANA_HOUSES = [2, 5, 9, 11]
PURE_KENDRAS = [4, 7, 10]
PURE_TRIKONAS = [5, 9]
PLANETS = ["sun", "moon", "mars", "mercury", "jupiter", "venus", "saturn"]


def sign_lord(sign):
    """The traditional ruler of a sign index (0 = Aries)."""
    return SIGN_RULERS[sign % 12]


def house_sign(asc_sign, house):
    """The sign on the given whole-sign house (1-12) from the Ascendant."""
    return (asc_sign + house - 1) % 12


def house_lord(asc_sign, house):
    """The lord of the given whole-sign house from the Ascendant."""
    return sign_lord(house_sign(asc_sign, house))


def house_from_asc(asc_sign, sign):
    """The whole-sign house (1-12) a sign falls in from the Ascendant."""
    return (sign - asc_sign) % 12 + 1


def aspects_sign(planet, planet_sign, target_sign):
    """Whether `planet` at `planet_sign` casts a graha drishti onto target_sign."""
    dist = (target_sign - planet_sign) % 12 + 1
    return dist in DRISHTI.get(planet, [7])


def parivartana(planet_a, sign_a, planet_b, sign_b):
    """Sign exchange: a sits in b's sign and b sits in a's sign."""
    return sign_lord(sign_a) == planet_b and sign_lord(sign_b) == planet_a


def association_type(planet_a, sign_a, planet_b, sign_b):
    """How two planets associate: 'conjunction' | 'exchange' | 'aspect' | None.
    Fixed priority so both ports agree."""
    if planet_a == planet_b:
        return None
    if sign_a == sign_b:
        return "conjunction"
    if parivartana(planet_a, sign_a, planet_b, sign_b):
        return "exchange"
    if aspects_sign(planet_a, sign_a, sign_b) and aspects_sign(planet_b, sign_b, sign_a):
        return "aspect"
    return None


def yogakarakas(asc_sign):
    """Planets that rule both a pure kendra (4/7/10) and a pure trikona (5/9)
    from the Ascendant -- the classic yogakarakas. Sorted."""
    out = []
    for p in PLANETS:
        ruled = {h for h in range(1, 13) if house_lord(asc_sign, h) == p}
        if ruled & set(PURE_KENDRAS) and ruled & set(PURE_TRIKONAS):
            out.append(p)
    return sorted(out)


def _lord_pair_yogas(asc_sign, signs, houses_a, houses_b):
    """Unique associated lord-pairs where one lord rules a house in houses_a and
    the other a house in houses_b. Returns sorted [{lords, via}]."""
    lords_a = sorted({house_lord(asc_sign, h) for h in houses_a})
    lords_b = sorted({house_lord(asc_sign, h) for h in houses_b})
    seen = {}
    for la in lords_a:
        for lb in lords_b:
            via = association_type(la, signs[la], lb, signs[lb])
            if via is None:
                continue
            pair = tuple(sorted([la, lb]))
            seen.setdefault(pair, via)
    return [{"lords": list(p), "via": v} for p, v in sorted(seen.items())]


def raja_yogas(signs, asc_sign):
    """Raja yogas: associations between a kendra lord and a trikona lord.
    `signs` maps each of the seven planets to its 0-based sign index."""
    return _lord_pair_yogas(asc_sign, signs, KENDRAS, TRIKONAS)


def dhana_yogas(signs, asc_sign):
    """Dhana yogas: associations between two wealth-house (2/5/9/11) lords."""
    return _lord_pair_yogas(asc_sign, signs, DHANA_HOUSES, DHANA_HOUSES)


def _signs_of(engine, natal_jd, lat, lon_east, zodiac):
    import math
    chart = engine.chart_at(natal_jd, lat, lon_east, zodiac=zodiac)
    asc_sign = math.floor(chart["angles"]["asc"] / 30.0) % 12
    signs = {p: math.floor(chart["bodies"][p]["lon"] / 30.0) % 12 for p in PLANETS}
    return signs, asc_sign


def raja_yogas_at(engine, natal_jd, lat, lon_east, zodiac="sidereal:lahiri"):
    """Raja yogas of a natal chart, with the chart's yogakarakas."""
    signs, asc_sign = _signs_of(engine, natal_jd, lat, lon_east, zodiac)
    return {"raja": raja_yogas(signs, asc_sign), "yogakarakas": yogakarakas(asc_sign)}


def dhana_yogas_at(engine, natal_jd, lat, lon_east, zodiac="sidereal:lahiri"):
    """Dhana yogas of a natal chart."""
    signs, asc_sign = _signs_of(engine, natal_jd, lat, lon_east, zodiac)
    return dhana_yogas(signs, asc_sign)

File: packages/caelus/src/rajayoga.ts
/**
 * astroengine rajayoga -- the lordship-and-aspect layer for Vedic yogas, and the
 * raja/dhana yogas built on it.
 *
 * Layers: house lordship (traditional ruler of each whole-sign house from the
 * Ascendant); graha drishti (every planet aspects the 7th sign; Mars also 4/8,
 * Jupiter 5/9, Saturn 3/10); association (conjunction, mutual aspect, or
 * parivartana exchange); then raja yoga (a kendra lord 1/4/7/10 associated with
 * a trikona lord 1/5/9), dhana yoga (two wealth-house 2/5/9/11 lords), and the
 * yogakaraka (a planet ruling both a pure kendra 4/7/10 and a pure trikona 5/9).
 * Definitions follow BPHS, validated against the named source in
 * `validate_jyotish`. Mirrors the Python reference (astroengine/rajayoga.py).
 */
import { Engine, BodyId, Zodiac } from "./chart.js";
import { SIGN_RULERS } from "./profections.js";

/** Graha drishti: the house-distances (1-based) each planet aspects. */
export const DRISHTI: Record<string, number[]> = {
  sun: [7], moon: [7], mercury: [7], venus: [7],
  mars: [4, 7, 8], jupiter: [5, 7, 9], saturn: [3, 7, 10],
};
export const KENDRAS = [1, 4, 7, 10];
export const TRIKONAS = [1, 5, 9];
export const DHANA_HOUSES = [2, 5, 9, 11];
const PURE_KENDRAS = [4, 7, 10];
const PURE_TRIKONAS = [5, 9];
const PLANETS: BodyId[] = ["sun", "moon", "mars", "mercury", "jupiter", "venus", "saturn"];

/** The traditional ruler of a sign index (0 = Aries). */
export function signLord(sign: number): string {
  return SIGN_RULERS[((sign % 12) + 12) % 12];
}

/** The sign on the given whole-sign house (1-12) from the Ascendant. */
export function houseSign(ascSign: number, house: number): number {
  return (ascSign + house - 1) % 12;
}

/** The lord of the given whole-sign house from the Ascendant. */
export function houseLord(ascSign: number, house: number): string {
  return signLord(houseSign(ascSign, house));
}

/** The whole-sign house (1-12) a sign falls in from the Ascendant. */
export function houseFromAsc(ascSign: number, sign: number): number {
  return ((sign - ascSign) % 12 + 12) % 12 + 1;
}

/** Whether `planet` at `planetSign` casts a graha drishti onto `targetSign`. */
export function aspectsSign(planet: string, planetSign: number, targetSign: number): boolean {
  const dist = ((targetSign - planetSign) % 12 + 12) % 12 + 1;
  return (DRISHTI[planet] ?? [7]).includes(dist);
}

/** Sign exchange: a sits in b's sign and b sits in a's sign. */
export function parivartana(planetA: string, signA: number, planetB: string, signB: number): boolean {
  return signLord(signA) === planetB && signLord(signB) === planetA;
}

/** How two planets associate: "conjunction" | "exchange" | "aspect" | null. */
export function associationType(
  planetA: string, signA: number, planetB: string, signB: number,
): string | null {
  if (planetA === planetB) return null;
  if (signA === signB) return "conjunction";
  if (parivartana(planetA, signA, planetB, signB)) return "exchange";
  if (aspectsSign(planetA, signA, signB) && aspectsSign(planetB, signB, signA)) return "aspect";
  return null;
}

/** Planets ruling both a pure kendra (4/7/10) and a pure trikona (5/9), sorted. */
export function yogakarakas(ascSign: number): string[] {
  const out: string[] = [];
  for (const p of PLANETS) {
    const ruled = new Set<number>();
    for (let h = 1; h <= 12; h++) if (houseLord(ascSign, h) === p) ruled.add(h);
    if (PURE_KENDRAS.some((h) => ruled.has(h)) && PURE_TRIKONAS.some((h) => ruled.has(h))) out.push(p);
  }
  return out.sort();
}

export interface LordPairYoga { lords: string[]; via: string; }

function lordPairYogas(
  ascSign: number, signs: Record<string, number>, housesA: number[], housesB: number[],
): LordPairYoga[] {
  const lordsA = [...new Set(housesA.map((h) => houseLord(ascSign, h)))].sort();
  const lordsB = [...new Set(housesB.map((h) => houseLord(ascSign, h)))].sort();
  const seen = new Map<string, string>();
  for (const la of lordsA) {
    for (const lb of lordsB) {
      const via = associationType(la, signs[la], lb, signs[lb]);
      if (via === null) continue;
      const pair = [la, lb].sort().join("|");
      if (!seen.has(pair)) seen.set(pair, via);
    }
  }
  return [...seen.entries()].sort((a, b) => a[0].localeCompare(b[0]))
    .map(([pair, via]) => ({ lords: pair.split("|"), via }));
}

/** Raja yogas: associations between a kendra lord and a trikona lord. */
export function rajaYogas(signs: Record<string, number>, ascSign: number): LordPairYoga[] {
  return lordPairYogas(ascSign, signs, KENDRAS, TRIKONAS);
}

/** Dhana yogas: associations between two wealth-house (2/5/9/11) lords. */
export function dhanaYogas(signs: Record<string, number>, ascSign: number): LordPairYoga[] {
  return lordPairYogas(ascSign, signs, DHANA_HOUSES, DHANA_HOUSES);
}

function signsOf(engine: Engine, natalJd: number, lat: number, lonEast: number, zodiac: Zodiac) {
  const chart = engine.chartAt(natalJd, lat, lonEast, { zodiac });
  const ascSign = Math.floor(chart.angles.asc / 30) % 12;
  const signs: Record<string, number> = {};
  for (const p of PLANETS) signs[p] = Math.floor(chart.bodies[p].lon / 30) % 12;
  return { signs, ascSign };
}

/** Raja yogas of a natal chart, with the chart's yogakarakas. */
export function rajaYogasAt(
  engine: Engine, natalJd: number, lat: number, lonEast: number, zodiac: Zodiac = "sidereal:lahiri",
): { raja: LordPairYoga[]; yogakarakas: string[] } {
  const { signs, ascSign } = signsOf(engine, natalJd, lat, lonEast, zodiac);
  return { raja: rajaYogas(signs, ascSign), yogakarakas: yogakarakas(ascSign) };
}

/** Dhana yogas of a natal chart. */
export function dhanaYogasAt(
  engine: Engine, natalJd: number, lat: number, lonEast: number, zodiac: Zodiac = "sidereal:lahiri",
): LordPairYoga[] {
  const { signs, ascSign } = signsOf(engine, natalJd, lat, lonEast, zodiac);
  return dhanaYogas(signs, ascSign);
}
